web/src/test/auto/headless/engine/osk/input/gestures/browser/subkeyPopup.tests.ts

```diff
@@ -0,0 +1,172 @@
+import { assert } from 'chai';
+import { JSDOM } from 'jsdom';
+import sinon from 'sinon';
+
+import { GesturePath, GestureSequence, GestureSource, GestureSourceSubview } from '@keymanapp/gesture-recognizer';
+
+import { ActiveSubKey } from '@keymanapp/keyboard-processor';
+import { DeviceSpec } from '../../../../../../../../../../common/web/utils/build/obj/deviceSpec.js';
+import OSKSubKey from '../../../../../../../../../build/engine/osk/obj/input/gestures/browser/oskSubKey.js';
+import OSKBaseKey from '../../../../../../../../../build/engine/osk/obj/keyboard-layout/oskBaseKey.js';
+import OSKRow from '../../../../../../../../../build/engine/osk/obj/keyboard-layout/oskRow.js';
+import { default as SubkeyPopup } from '../../../../../../../../../build/engine/osk/obj/input/gestures/browser/subkeyPopup.js'
+import { link } from '../../../../../../../../../build/engine/osk/obj/keyElement.js'
```

Comment:
I would merge most of these, where possible, into line 14's `import`.

The tests themselves should be fine here, since this is the "headless" section.  The distinction matters quite significantly when code being tested is bundled for testing in-browser, though!

Reply:
Done where possible. Most of these imports are not exported in `keyman/engine/osk` though.

Reply:
This is a .ts, why are we not importing from the source .ts files (as .js) rather than the build artifacts?

Reply:
It would be wise to add them as exports, even if only for availability when testing.

Alternatively, we could create a test-index module (as an export point) to allow access to internals for testing mode, which could give us something like:

```typescript
import OSKSubKey from 'keyman/engine/osk/internals';
```

... all without exposing them with the primary export.  The _larger_ benefit, of course, is having a common, relative-path anchor point that can be updated in a single location if necessary.

----

So, how to do that?

For reference, here are some excerpts from Web's package.json that were used to establish `'keyman/engine/osk'` as a viable import path:

https://github.com/keymanapp/keyman/blob/bb54a9aa0a5916109884e156142c78d8883c205f/web/package.json#L1-L4

https://github.com/keymanapp/keyman/blob/bb54a9aa0a5916109884e156142c78d8883c205f/web/package.json#L61-L65

We can just replicate the second excerpt and change it a bit to point to a test-index.ts via the name `./engine/osk/internals`:

```json
"./engine/osk/internals": { 
  "types": "./build/engine/osk/obj/test-index.d.ts", 
  "import": "./build/engine/osk/obj/test-index.js" 
} 
```

This would then establish the prettier import style.  There's no need for the `es6-bundling` variant in the export map, since we don't bundle the test index.  If we did want it anyway...

```
"es6-bundling": "./src/engine/osk/src/test-index.ts"
```